### PR TITLE
Systemd improvements

### LINKED ISSUE #664: Pass correct OS PID to systemd-notify
See the discussion in https://github.com/rabbitmq/rabbitmq-server/pull/573.

### PR DESCRIPTION
**Pull request by binarin**
After some investigation I think that even with `sd_notify` MAINPID
detection didn't work 100% percent correctly.

Foreground broker was started through a shell process that was
converting signals into graceful shutdown commands. And it's pid of that
proxy that was to be communicated to systemd. Automatic pid inference
worked in case epmd was already running on machine, otherwise it was epmd
who was chosen as main process.

This commit introduces new environment variable to prevent creation of
such a proxy process. And this var should be specified in unit
file (example added to docs/). New variable is needed due to following
reasons:
- No way to detect whether we are running under systemd or not
- I was thinking about adding command-line switch instead, but there is
  no way to work with "$@" in portable shell script.

Another improvement is writing message to stderr about graceful
shutdown, just to be sure that shutdown was indeed graceful. In case of
systemd it will automatically appear in the journal.

Fixes #664

### CODE REVIEW

**Comment by michaelklishin** [comment]
@binarin this no longer merges cleanly, can you please rebase? Thank you.

File: docs/rabbitmq-server.service.example [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+[Unit]
+Description=RabbitMQ broker
+After=syslog.target network.target
+
+[Service]
+Type=notify
+User=rabbitmq
+Group=rabbitmq
+NotifyAccess=all
+TimeoutStartSec=3600
```
Comment (dumbbell) [inline docs/rabbitmq-server.service.example]: I see you added a

File: src/rabbit.erl [hunk 2]
```diff
@@ -284,16 +284,94 @@ broker_start() ->
...
+            sd_notify_socat(Unit);
+        _ ->
+            false
+    end.
+
+sd_notify_socat(Unit) ->
+    case catch open_port(
+                 {spawn_executable, os:find_executable("socat")},
+                 [{args, ["unix-sendto:" ++ os:getenv("NOTIFY_SOCKET"), "STDIO"]},
+                  exit_status, use_stdio, out]) of
```
Comment (dumbbell) [inline src/rabbit.erl]: You never call `port_close/1`, how is socat(1) termianted?

File: src/rabbit.erl [hunk 3]
```diff
@@ -284,16 +284,94 @@ broker_start() ->
...
+sd_notify_socat() ->
+    case sd_current_unit() of
+        {ok, Unit} ->
+            io:format(standard_error, "systemd unit for activation check: \"~s\"~n", [Unit]),
+            sd_notify_socat(Unit);
+        _ ->
+            false
+    end.
+
+sd_notify_socat(Unit) ->
```
Comment (michaelklishin) [inline src/rabbit.erl]: Please add a comment linking to #664. It's not obvious why we have to use `socat` and other things that are going on here.

File: docs/rabbitmq-server.service.example (hunk 1 above)
Comment (binarin) [inline docs/rabbitmq-server.service.example]: Recently I'm very suspicious of `infinity` timeouts. I don't remember who said something like: 
> When I see a timeouf of `inifinity`, I suggest to replace it with some insane value like 30 years. And if the reply to this is that 30 years doesn't make sense, I reply that infinity also makes no sense.

With `infinity` timeout some problem with startup could go unnoticed for a long period of time. The question is what is the sane timeout value - in my opininon it's no more than several hours.

File: src/rabbit.erl (hunk 2 above)
Comment (binarin) [inline src/rabbit.erl]: I saw that it terminated and never asked that question. This happens unintentionally due to controlling process death. I'm going to make it explicit.

File: docs/rabbitmq-server.service.example (hunk 1 above)
Comment (dumbbell) [inline docs/rabbitmq-server.service.example]: I discussed this with @michaelklishin and he agrees with you, so we'll keep this timeout of one hour.

**Comment by dumbbell** [comment]
Thank you @binarin!

**Pull request merged**